scripts: add a plandiff script

asubiotto opened the pull request:
This script consumes a path to a cockroach binary, a file of queries, and the
ports for two different cockroach clusters. It runs an explain on each query
against both clusters and checks for differences.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47568)
<!-- Reviewable:end -->

asubiotto commented:
Some other additions that might be interesting:
- Adding a `--database` flag or maybe swapping ports out for a pgurl
- Allowing an optional argument to specify the explain type desired (e.g. EXPLAIN ANALYZE)

yuzefovich reviewed (commented):
I don't have pretty much any experience writing scripts, but it LGTM.

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47568#-:-M53jwTzBDmaTEGaWaDZ:bh2d6tr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[scripts/plandiff, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47568#-M53iyKv4ewBIFqsQyJn:-M53iyKw2ZBEatGx66fU:bec9l1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/125c48e51b5ba5824d8d8787fc0893431d9ae830/scripts/plandiff#L47)):*
> ```
>     tmpfile=$(mktemp)
>     if ! diff "$dest/query$i.old" "$dest/query$i.new" > $tmpfile; then
>         echo "found difference in plans for query $i"
> ```

I think it'd be useful to print out the query itself as well, probably using `sqlfum.pt`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47568)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[scripts/plandiff, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47568#-M53iyKv4ewBIFqsQyJn:-M53k7Gp6vtVSpre8452:bi0h7lr) ([raw file](https://github.com/cockroachdb/cockroach/blob/125c48e51b5ba5824d8d8787fc0893431d9ae830/scripts/plandiff#L47)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think it'd be useful to print out the query itself as well, probably using `sqlfum.pt`.
</blockquote></details>

Also adding a few blank lines between different plans would be nice too.


<!-- Sent from Reviewable.io -->

asubiotto commented:
Although it'd be nice, adding sqlfum.pt would introduce a tool dependency which seems unnecessary. Updated with pgurls and optional explain variant. I added a line between diff outputs  but I'm not sure whether it changes much.

yuzefovich reviewed (commented):
WDYT about printing the query as is then when the plans differ?

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47568#-:-M57nGC2EtLAk_DDKnn5:b-jdtpbj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)



<!-- Sent from Reviewable.io -->

jordanlewis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, seems like it could be useful

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47568#-:-M5AU9MQ4GNyzC75ojDX:bfarvmk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)



<!-- Sent from Reviewable.io -->

Commit:
scripts: add a plandiff script

This script consumes a path to a cockroach binary, a file of queries, and the
ports for two different cockroach clusters. It runs an explain on each query
against both clusters and checks for differences.

Release note: None

asubiotto commented:
Done.

bors r=yuzefovich,jordanlewis
